File: Packages/NMP/Sources/NMP/Query.swift
// The pull-based Rust-handle -> Swift-`AsyncSequence` adapter (#680). This is
// the ONLY place `NMP` folds observation frames into delivered state.

import Foundation
import NMPFFI

/// A live, detachable query (`nmp_engine`'s read noun). `NMPQuery` is the
/// PRIMARY read handle -- iterate it directly with `for try await`; there is
/// no container or provider object required around it (M4 plan §7's canary).
///
/// Each element is the full current snapshot (`RowBatch`), never a bare
/// delta. How that snapshot is produced derives from the observation's
/// boundedness (#485): an UNBOUNDED query is delivered as exact rebased
/// deltas that the iterator folds into its accumulated state (the engine
/// mailbox conflates intermediate reducer emits for a slow consumer); a
/// WINDOWED query is delivered as authoritative bounded snapshots that
/// replace the state wholesale, each carrying the window's `WindowLoad` fact.
///
/// The sequence is THROWING: it ends normally (`nil`) when the engine tears
/// the subscription down, and surfaces `NMPError.concurrentNext` if two
/// iterators pull the same handle at once (the handle is single-consumer,
/// #680). The direct iterator cadence-limits complete snapshots (#17) so a
/// tight `for try await` loop during historical replay cannot peg the main
/// thread; values produced while it waits conflate in the native mailbox.
///
/// Demand teardown is ITERATOR-OWNED: dropping the iterator on normal scope
/// exit (including `break`) cancels the handle and releases its sequence claim.
/// Task cancellation is forwarded synchronously to the native handle so a
/// parked `next()` wakes; Swift cancellation does not cross UniFFI by itself.
public struct NMPQuery: AsyncSequence, Sendable {
    public typealias Element = RowBatch

    private let handle: NmpRowStream
    private let iteratorGate = NMPPullIteratorGate()

    init(engine: NmpEngineProtocol, filter: FfiFilter, window: FfiWindow?) throws {
        self.handle = try nmpRethrowing {
            try engine.observe(query: filter, window: window)
        }
    }

    /// #107: the explicit-`FfiDemand` entry point -- same handle/coalescing
    /// shape as the `FfiFilter` initializer above, just a different
    /// `NmpEngineProtocol` verb underneath.
    init(engine: NmpEngineProtocol, demand: FfiDemand, window: FfiWindow?) throws {
        self.handle = try nmpRethrowing {
            try engine.observeDemand(query: demand, window: window)
        }
    }

    public func makeAsyncIterator() -> Iterator {
        let accumulator = RowAccumulator()
        let core = NMPPullIteratorCore(
            handle: handle,
            iteratorGate: iteratorGate,
            throttle: true
        ) { frame in accumulator.fold(frame) }
        return Iterator(core: core)
    }

    public struct Iterator: AsyncIteratorProtocol {
        let core: NMPPullIteratorCore<NmpRowStream, RowBatch>

        public mutating func next() async throws -> RowBatch? {
            try await core.next()
        }
    }

    /// Windowed observations only: monotonically raise this query's window
    /// row target to at least `atLeast`, clamped to the window's declared
    /// `max`. Growth is DECLARATIVE by design -- no continuation token to
    /// round-trip, so there is nothing to go stale and nothing to misuse;
    /// the call is idempotent, and a value at or below the current target
    /// is simply a no-op. Outcomes arrive in-band as `WindowLoad` facts on
    /// delivered batches (`RowBatch.load`) -- including `.atBound(max:)`,
    /// which is a delivered fact, never a thrown error.
    ///
    /// Throws only the synchronous refusals: `NMPRequestRowsError`
    /// (`.unwindowed` on a query opened without a window, `.engineClosed`,
    /// `.storeUnavailable`).
    public func requestRows(atLeast: UInt64) throws {
        do {
            try handle.requestRows(atLeast: atLeast)
        } catch let error as FfiRequestRowsError {
            throw NMPRequestRowsError(error)
        }
    }

    /// Withdraw the subscription now rather than waiting for the last
    /// reference to be released. Safe to call more than once; safe to never
    /// call at all.
    public func cancel() {
        handle.cancel()
    }
}

/// The unbounded/windowed delta-fold accumulator, moved out of the deleted
/// `RowObserver` bridge and into the iterator's per-frame mapping (#680).
/// `internal` (not `private`) so `@testable import NMP` can drive `fold`
/// directly for the accumulation/replacement falsifiers (#105's `SourcesGrew`
/// replace-in-place proof; #485's windowed-snapshot replacement proof).
///
/// ONE fold, two frame shapes, chosen by the engine from the observation's
/// boundedness (#485):
///
/// - `frame.window == nil` (unbounded): `frame.deltas` is the exact
///   transition rebased onto the last delivered Rust frame. Folding every
///   delivered transition keeps the accumulated state exact.
/// - `frame.window != nil` (windowed): `frame.window!.rows` is the complete
///   authoritative bounded set and REPLACES the state wholesale -- windowed
///   frames conflate to latest-state on the Rust side, so `frame.deltas` is
///   always empty here.
///
/// Only ever touched from the single pump task that owns its stream, so no
/// lock is needed (the old bridge locked because a callback thread raced the
/// consumer; there is no callback thread anymore).
final class RowAccumulator: @unchecked Sendable {
    // Insertion-ordered accumulation for the unbounded mode: `order` tracks
    // arrival order, `byId` the current value for each still-live row. For
    // the windowed mode both are replaced from each authoritative frame
    // (canonical newest-first order). NMP does mechanics only (retain what the
    // engine says is live) -- ordering/rendering policy is an app concern.
    private var order: [String] = []
    private var byId: [String: Row] = [:]

    func fold(_ frame: FfiFrame) -> RowBatch {
        if let window = frame.window {
            // #485: an authoritative bounded snapshot -- replace, never fold.
            // `frame.deltas` is empty by contract for windowed frames (rows
            // never cross the FFI twice).
            let rows = window.rows.map(Row.init)
            order = rows.map(\.id)
            byId = Dictionary(uniqueKeysWithValues: rows.map { ($0.id, $0) })
            return RowBatch(
                rows: rows,
                evidence: AcquisitionEvidence(frame.evidence),
                load: WindowLoad(window.load)
            )
        }
        for delta in frame.deltas {
            switch delta {
            case .added(let ffiRow):
                let row = Row(ffiRow)
                if byId[row.id] == nil {
                    order.append(row.id)
                }
                byId[row.id] = row
            case .sourcesGrew(let id, let sources):
                // #105: the SAME row already matched; only its relay-provenance
                // set grew. Replace it in place -- `order` is untouched, this is
                // never an insertion.
                if let existing = byId[id] {
                    byId[id] = existing.withSources(sources)
                }
            case .removed(let id):
                if byId.removeValue(forKey: id) != nil {
                    order.removeAll { $0 == id }
                }
            }
        }
        let snapshot = order.compactMap { byId[$0] }
        return RowBatch(
            rows: snapshot,
            evidence: AcquisitionEvidence(frame.evidence),
            load: nil
        )
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Query.kt
// The Rust-handle -> Kotlin-`Flow` bridge -- the Kotlin counterpart of
// Query.swift's `NMPQuery`/`RowBridge`. This is the heart of the ergonomic
// layer: the ONLY place `com.nmp.sdk` folds the exact `FfiFrame` deltas into
// the accumulated snapshot a collector sees.
//
// PULL MODEL (#680): the engine owns a single-slot latest-state mailbox per
// observation; the wrapper synchronously claims an `NmpRowPull`, awaits its
// frame, and commits before folding the delta into the running snapshot.
// `null` from `receive()` is the terminal signal (cancel / engine shutdown /
// producer drop). There is NO callback observer, NO dedicated drain thread,
// and NO native-task capacity concept anymore.
//
// TEARDOWN is COLLECTION-SCOPE-TIED: `handle.cancel()` runs in a `finally`,
// which fires the moment the collecting coroutine is cancelled or the flow
// completes. Cancelling the collecting coroutine drops the in-flight Rust
// `receive()` future and aborts the still-owned pull ticket, restoring any
// delta that reached Rust READY before generated Kotlin completion retrieved
// it. This is the deterministic teardown #46's bounded-latest-state contract
// needs -- not the JVM `Cleaner`, which only runs once GC actually collects.
//
// CONFLATION lives in the engine mailbox now, so there is deliberately NO
// `.conflate()` on this side: a slow collector simply does not call `next()`
// again until it is ready, and the engine folds intermediate reducer emits
// into the single retained slot (exact deltas are rebased onto the last
// delivered frame). Backpressure is expressed by not pulling.
package com.nmp.sdk

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import uniffi.nmp_ffi.FfiFrame
import uniffi.nmp_ffi.FfiRowPullException
import uniffi.nmp_ffi.FfiRowDelta
import uniffi.nmp_ffi.NmpEngineInterface
import uniffi.nmp_ffi.NmpRowStream

/** A live, detachable query (`nmp_engine`'s read noun). The returned `Flow`
 * is the PRIMARY read handle -- `collect` it directly; there is no
 * container or provider object required around it (mirrors `NMPQuery`'s own
 * doc / the M4 plan §7 canary).
 *
 * Each element is the full accumulated snapshot (`RowBatch`), never a bare
 * delta: the pull loop folds `Added`/`Removed`/`SourcesGrew` deltas
 * internally so a collector never has to. The cold `Flow` opens a fresh
 * single-consumer stream on each collection and withdraws it (via
 * `handle.cancel()`) the moment collection ends -- see this file's header
 * for why that, not UniFFI's generated `Cleaner`, is the correct mapping. */
fun observeQuery(engine: NmpEngineInterface, filter: NMPFilter): Flow<RowBatch> =
    rowFlow { nmpRethrowing { engine.observe(filter.toFfi(), null) } }

/** #1108: the explicit-`NMPLiveQuery` entry point -- the constructor to
 * reach for once [observeQuery]'s implicit `AuthorOutboxes`/`Public` default
 * isn't enough: declaring `NMPSourceAuthority.Pinned` wire authority, a
 * non-default `NMPAccessContext`, a non-`Agnostic` `NMPCacheMode`, SEVERAL
 * independent demand branches, or a bound on their merged row union. Every
 * branch is observed through this ONE stream. Same pull-loop/accumulation/
 * teardown shape as the `NMPFilter` overload above. */
fun observeQuery(engine: NmpEngineInterface, query: NMPLiveQuery): Flow<RowBatch> =
    rowFlow { nmpRethrowing { engine.observeQuery(query.toFfi(), null) } }

/** Shared pull loop for the unbounded (delta-folding) row observations.
 * `open` is the ONE difference between the `NMPFilter` and `NMPDemand` entry
 * points (which `NmpEngineInterface` verb actually opens the subscription),
 * run lazily per collection so the `Flow` stays cold.
 *
 * `internal`, not `private`, for the same reason as [nextCommittedRowFrame]
 * and [applyRowDelta] below: #1192 exercises this exact commit-then-fold-
 * then-emit ordering against a scripted [NmpRowStream] with a real cancelled
 * `Job`, which needs this loop directly rather than a live engine. */
internal fun rowFlow(open: () -> NmpRowStream): Flow<RowBatch> =
    flow {
        val handle = open()
        try {
            // Insertion-ordered accumulation: `order` tracks arrival order,
            // `byId` the current value for each still-live row -- same shape
            // as Swift's `RowBridge`. `com.nmp.sdk` does mechanics only
            // (accumulate what the engine says is live); ordering/rendering
            // policy is an app concern (feed doctrine), not this loop's.
            val order = mutableListOf<String>()
            val byId = mutableMapOf<String, Row>()
            while (true) {
                // These observations are opened with `window = null`, so
                // every delivered `frame.window` is null and `frame.deltas`
                // is the exact transition rebased onto the last delivered
                // Rust frame -- fold it. (Windowed observations take the
                // other arm of the one frame vocabulary: authoritative
                // snapshots, no folding -- see Window.kt.)
                val frame = nextCommittedRowFrame(handle) ?: break
                for (delta in frame.deltas) applyRowDelta(order, byId, delta)
                val snapshot = order.mapNotNull { byId[it] }
                emit(RowBatch(snapshot, frame.evidence.map { AcquisitionEvidence.from(it) }))
            }
        } finally {
            handle.cancel()
        }
    }

/**
 * #762's private two-phase foreign-delivery bridge.
 *
 * [NmpRowStream.beginNext] creates the exclusive native ticket before Kotlin
 * enters UniFFI's cancellable READY/complete split. If cancellation wins
 * before generated `receive()` retrieves the Rust result, this function still
 * owns [pull] and aborts it in `finally`, restoring the retained delta for the
 * next ticket. Once `receive()` returns, `commit()` is synchronous and runs
 * before mapping, folding, emitting, or any other cancellation point.
 */
internal suspend fun nextCommittedRowFrame(handle: NmpRowStream): FfiFrame? {
    val pull =
        try {
            handle.beginNext()
        } catch (error: FfiRowPullException) {
            return when (error) {
                is FfiRowPullException.Closed -> null
                else -> throw NMPError.ConcurrentNext
            }
        }
    var committed = false
    try {
        val frame = pull.receive()
        pull.commit()
        committed = true
        return frame
    } catch (error: FfiRowPullException) {
        return when (error) {
            is FfiRowPullException.Closed,
            is FfiRowPullException.Aborted,
            -> null
            else -> throw NMPError.ConcurrentNext
        }
    } finally {
        try {
            if (!committed) pull.abort()
        } finally {
            pull.close()
        }
    }
}

/**
 * The accumulator's per-delta step, extracted as a pure function so it is
 * directly unit-testable (#105's `SourcesGrew` replace-in-place proof)
 * without driving the coroutine/`Flow` machinery around it. Mutates
 * `order`/`byId` in place -- identical semantics to `RowBridge.onFrame`'s
 * Swift counterpart.
 */
internal fun applyRowDelta(
    order: MutableList<String>,
    byId: MutableMap<String, Row>,
    delta: FfiRowDelta,
) {
    when (delta) {
        is FfiRowDelta.Added -> {
            val row = Row.from(delta.row)
            if (!byId.containsKey(row.id)) order.add(row.id)
            byId[row.id] = row
        }
        is FfiRowDelta.SourcesGrew -> {
            // #105: the SAME row already matched; only its relay-provenance
            // set grew. Replace in place -- `order` is untouched, this is
            // never an insertion.
            byId[delta.id]?.let { existing ->
                byId[delta.id] = existing.copy(sources = delta.sources)
            }
        }
        is FfiRowDelta.Removed -> {
            if (byId.remove(delta.id) != null) order.remove(delta.id)
        }
    }
}
